h11/_receivebuffer.py

```diff
@@ -79,34 +87,57 @@ def maybe_extract_at_most(self, count):
         self._start += len(out)
         return out
 
-    def maybe_extract_until_next(self, needle):
+    def maybe_extract_until_next(self, needle_regex, max_needle_length):
         # Returns extracted bytes on success (advancing offset), or None on
         # failure
-        if self._looked_for == needle:
-            search_start = max(self._start, self._looked_at - len(needle) + 1)
+        if self._looked_for_regex == needle_regex:
+            looked_at = max(self._start, self._looked_at - max_needle_length)
         else:
-            search_start = self._start
-        offset = self._data.find(needle, search_start)
-        if offset == -1:
+            looked_at = self._start
+            self._looked_for_regex = needle_regex
+
+        delimiter_match = next(
+            self._looked_for_regex.finditer(self._data, looked_at), None
+        )
+
+        if delimiter_match is None:
             self._looked_at = len(self._data)
-            self._looked_for = needle
             return None
-        new_start = offset + len(needle)
-        out = self._data[self._start : new_start]
-        self._start = new_start
+
+        _, end = delimiter_match.span(0)
+
+        out = self._data[self._start : end]
+
+        self._start = end
+
         return out
 
+    def _get_fields_delimiter(self, data, lines_delimiter_regex):
+        delimiter_match = next(lines_delimiter_regex.finditer(data), None)
+
+        if delimiter_match is not None:
+            begin, end = delimiter_match.span(0)
+            result = data[begin:end]
+        else:
+            result = b"\r\n"
+
+        return bytes(result)
+
     # HTTP/1.1 has a number of constructs where you keep reading lines until
     # you see a blank one. This does that, and then returns the lines.
     def maybe_extract_lines(self):
-        if self._data[self._start : self._start + 2] == b"\r\n":
-            self._start += 2
+        start_chunk = self._data[self._start : self._start + 2]
+        if start_chunk in [b"\r\n", b"\n"]:
+            self._start += len(start_chunk)
             return []
         else:
-            data = self.maybe_extract_until_next(b"\r\n\r\n")
+            data = self.maybe_extract_until_next(blank_line_delimiter_regex, 3)
```

Comment:
I find it odd that `blank_line_delimiter_regex` is not `\r?\n\r?\n`, although I think the logic is find like this. If you don't think it is odd, then don't worry about it.

Reply:
Should it be `(\r\n\r\n|\n\n)` to be precise, on the basis that senders either always use spec-compliant line endings, or never do, rather than ever using a mix?

The `\r?\n\r?\n` version would match `"\r\n\n"` and `"\n\r\n"`.

Reply:
Well, I took `"\n\r?\n"` from this https://github.com/python-hyper/h11/issues/7#issuecomment-722640685 , but I cannot disagree that it might be a bit confusing.

I changed regex to `"(\r\n\r\n|\n\n)"` , it still should fix the original problem with a server, delimiting headers and body using double-lf